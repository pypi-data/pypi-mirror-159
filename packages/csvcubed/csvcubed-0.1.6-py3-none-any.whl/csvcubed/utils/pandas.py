"""
Pandas Utils
------------

This file provides additional utilities for pandas typoe commands
"""
import logging
from typing import Tuple, List

import pandas as pd

from pathlib import Path

from csvcubed.models.validationerror import ValidationError

from csvcubed.models.cube.validationerrors import DuplicateColumnTitleError

specified_na_values = {
    "",
}

logger = logging.getLogger(__name__)

def read_csv(csv: Path, **kwargs) -> Tuple[pd.DataFrame, List[ValidationError]]:
    """
    :returns: a tuple of
        pd.DataFrame without the default na values being changes into NaN
        list of ValidationExceptions
    """
    df = pd.read_csv(
        csv, keep_default_na=False, na_values=specified_na_values, **kwargs
    )
    if not isinstance(df, pd.DataFrame):
        logger.debug("Expected a pandas dataframe when reading from CSV, value was %s", df)
        raise ValueError(
            f"Expected a pandas dataframe when reading from CSV, value was {type(df)}"
        )


    # Read first row as values rather than headers, so we can check for duplicate column titles
    col_title_counts = pd.read_csv(csv, header=None, nrows=1).iloc[0, :].value_counts()  # type: ignore
    duplicate_titles = list(col_title_counts[col_title_counts > 1].keys())
    if any(duplicate_titles):
        return df, [
            DuplicateColumnTitleError(csv_column_title=dupe_title)
            for dupe_title in duplicate_titles
        ]

    return df, []
