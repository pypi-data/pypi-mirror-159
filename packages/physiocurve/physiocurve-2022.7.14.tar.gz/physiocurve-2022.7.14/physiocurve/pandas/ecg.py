from physiocurve.common import estimate_samplerate
from physiocurve.ecg import ECG as ECGNp


class ECG(ECGNp):
    def __init__(self, series, samplerate=None):
        self._series = series
        if samplerate is None:
            samplerate = estimate_samplerate(series)

        super().__init__(series.to_numpy(), samplerate)

    @property
    def idxpwave(self):
        return self._series.index[self.argpwave]

    @property
    def idxrwave(self):
        return self._series.index[self.argrwave]

    @property
    def idxtwave(self):
        return self._series.index[self.argtwave]

    @property
    def pwaves(self):
        return self._series.iloc[self.argpwave]

    @property
    def rwaves(self):
        return self._series.iloc[self.argrwave]

    @property
    def twaves(self):
        return self._series.iloc[self.argtwave]
