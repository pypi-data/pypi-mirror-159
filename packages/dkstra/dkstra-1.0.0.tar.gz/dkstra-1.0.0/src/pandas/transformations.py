import datetime
import logging
import os
from typing import List, Dict

import numpy as np
import pandas as pd


def series_to_pandas(series_in: pd.Series) -> pd.DataFrame:
    """
    Transforms series to single row pandas
    Parameters
    ----------
    series_in

    Returns
    -------

    """
    return pd.DataFrame([series_in.values], columns=series_in.index)


def apply_function_to_multiple_keys(input_dict: dict, func, keys_to_use: list) -> dict:
    """
    Applies the defined function to multiple keys in the dict.
    Modifies the input dict
    Parameters
    ----------
    input_dict
    func
    keys_to_use
    Returns
    -------

    """
    for in_key in keys_to_use:
        input_dict[in_key] = func(input_dict[in_key])

    return input_dict


def dict_to_series(input_dict: dict) -> pd.Series:
    """
    Transforms a dict to a pdSeries, where the index will be the keys, and the values the row
    Parameters
    ----------
    input_dict

    Returns
    -------

    """
    return pd.Series(index=input_dict.keys(), data=list(input_dict.values()))


def dict_to_pandas(input_dict: dict, index: list = None) -> pd.DataFrame:
    """
    Transforms a dict to a pandas dataframe, where the index will be the keys, and the values the row
    Parameters
    ----------
    input_dict
    index

    Returns
    -------

    """
    return pd.DataFrame(
        [input_dict.values()], columns=list(input_dict.keys()), index=[index]
    )


def pandas_to_dict(df: pd.DataFrame) -> dict:
    """
    Transforms a single dataframe df into a dict with one pair key-value per column of the dataframe
    Parameters
    ----------
    df

    Returns
    -------

    """
    return df.to_dict(orient="records")[0]


def append_string_to_columns(df: pd.DataFrame, string_to_concat: str) -> pd.DataFrame:
    """

    Parameters
    ----------
    df
    string_to_concat

    Returns
    -------

    """
    df.columns = string_to_concat + df.columns
    return df


def select_and_rename(df: pd.DataFrame, mapping: dict) -> pd.DataFrame:
    """
    Selects columns of a dataframe and renames them
    Parameters
    ----------
    df
    mapping

    Returns
    -------

    """
    return df[mapping.keys()].rename(columns=mapping)


def three_sigma_fill_na(df: pd.DataFrame, col: str):
    upper = df[col].mean() + 3 * df[col].std()
    lower = df[col].mean() - 3 * df[col].std()

    df.loc[df[col] > upper, col] = np.nan
    df.loc[df[col] < lower, col] = np.nan

    return df


def df_threshold_fill_na(df: pd.DataFrame, col: str, lower=None, upper=None):
    df = df.copy()
    if upper is not None:
        df.loc[df[col] > upper, col] = np.nan
    if lower is not None:
        df.loc[df[col] < lower, col] = np.nan

    return df


def ffill_on_value(pd_series: pd.Series, val_cond: float) -> pd.Series:
    """

    Parameters
    ----------
    pd_series
    val_cond

    Returns
    -------

    """
    pd_series.loc[pd_series == val_cond] = np.nan
    pd_series = pd_series.ffill()
    pd_series = pd_series.fillna(val_cond)
    return pd_series


def merge_dataframes_dict(
    df_skeleton: pd.DataFrame,
    dict_df: Dict[str, pd.DataFrame],
    merge_settings: Dict[str, Dict[str, str]],
    ordered_keys: List[str],
) -> pd.DataFrame:
    """
    Merges a dict containing multiple dataframes and a set of settings
    Parameters
    ----------
    dict_df
    df_skeleton
    merge_settings
    ordered_keys

    Returns
    -------

    """

    df_skeleton = df_skeleton.copy()
    for key in ordered_keys:
        df_skeleton = pd.merge(df_skeleton, dict_df[key], **merge_settings[key])
    return df_skeleton


def calculate_weighted_average(value: pd.Series, weight: pd.Series, group: pd.Series) -> pd.Series:
    weighted_value = value * weight
    g_weighted_value = weighted_value.groupby(group).transform(sum)
    g_weight = weight.groupby(group).transform(sum)
    return g_weighted_value / g_weight


def merge_multiple_dataframes(
        df_skeleton: pd.DataFrame,
        list_df: List[pd.DataFrame],
        how: str, on:
        List[str]) -> pd.DataFrame:
    df = df_skeleton.copy()
    for df_to_add in list_df:
        df = pd.merge(df, df_to_add, how, on)
    return df


def round_all_columns(df: pd.DataFrame, n_round: int = 2) -> pd.DataFrame:
    for col in df.columns:
        if pd.api.types.is_numeric_dtype(df[col]) and not pd.api.types.is_bool_dtype(
            df[col]
        ):
            df[col] = np.round(df[col], n_round)
    return df


def explode_df(df: pd.DataFrame, list_df: List[pd.DataFrame]) -> pd.DataFrame:
    for df2 in list_df:
        df = cross_join(df, df2)
    return df


def cross_join(df1: pd.DataFrame, df2: pd.DataFrame) -> pd.DataFrame:
    """
    Cross joins two dataframes
    Parameters
    ----------
    df1
    df2

    Returns
    -------

    """
    df1 = df1.copy()
    df2 = df2.copy()
    df1["key"] = 0
    df2["key"] = 0
    df = df1.merge(df2, on="key", how="outer")
    df = df.drop(columns="key")
    return df

