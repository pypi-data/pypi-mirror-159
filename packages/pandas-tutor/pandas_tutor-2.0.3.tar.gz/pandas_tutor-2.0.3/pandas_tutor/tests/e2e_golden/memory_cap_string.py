import pandas as pd

# this uses about 2.5 MB
big = pd.DataFrame({'text': ['hello world' * 120_000]})

big.head()
