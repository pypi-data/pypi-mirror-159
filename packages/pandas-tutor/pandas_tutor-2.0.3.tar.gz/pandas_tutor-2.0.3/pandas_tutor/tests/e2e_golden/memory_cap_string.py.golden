{
  "code": "import pandas as pd\n\n# this uses about 2.5 MB\nbig = pd.DataFrame({'text': ['hello world' * 120_000]})\n\nbig.head()\n",
  "explanation": [
    {
      "type": "RuntimeErrorInChain",
      "code_step": "big.head()",
      "message": "Your total data uses 2.52 MB of memory, which exceeds the maximum of 1.00 MB that this tool supports.",
      "fragment": {
        "start": {
          "line": 0,
          "ch": 3
        },
        "end": {
          "line": 0,
          "ch": 10
        }
      }
    }
  ]
}